Slight refactoring and documentation of compression (#739)
FAIRmat-NFDI/pynxtools: docs/learn/pynxtools/compression.md

# Using compression with HDF5 

## Approach

> Data compression covers methods to effectively reduce the size of a dataset or portions of it. Lossless and lossy methods are distinguished. There is an entire research field working on developing methods and implementations for each method. Given that `pynxtools` writes its content to HDF5 files, we have decided to use the functionalities that this library provides to motivate users to use compression as an optional feature, at an optionally individual dataset level. To preserve the original data, we decided to support only with lossless compression algorithms. We also decided to currently not compress strings and scalar dataset.

Reviewer: ```suggestion
Data compression covers methods to effectively reduce the size of a dataset or portions of it. Lossless and lossy methods are distinguished. There is an entire research field working on developing methods and implementations for each method. Given that `pynxtools` writes its content to HDF5 files, we have decided to use the functionalities that this library provides to motivate users to use compression as an optional feature, at an optionally individual dataset level. To preserve the original data, we decided to only support lossless compression algorithms. We also decided to not compress strings and scalar dataset.
```
removing currently, if we change that, we can always update here

Specifically, we use the build-in [gzip (`deflate`) compression filter](https://support.hdfgroup.org/documentation/hdf5-docs/hdf5_topics/UsingCompressionInHDF5.html) due to its wide support across the most frequently used programming languages. Users should be aware that using `deflate` instead of more modern algorithms has the trade-off that there is as of now no efficient multi-threaded implementation of this compression filter within the HDF5 library. Therefore, compression can take a substantial amount of the total execution time within the `dataconverter` HDF5 file writing part.

## How to use compression

Developers of plugins for `pynxtools` instruct the writing of data via adding variables such as numpy or xarrays like the following example

```
array = np.zeros((1000,), np.float64)
```

to specific places in the `template` dictionary object that the `dataconverter` provides:

```
template["/ENTRY[entry1]/numpy_array"] = array
```

> For such an instruction, the `dataconverter` creates an HDF5 dataset instance that uses the so-called contiguous data storage layout. This dataset is stored uncompressed.

Author: Suggestion taken

As an alternative, compression can be instructed via a slight modification of the previous example:

```
template["/ENTRY[entry1]/array"] = {
  "compress": array,
  "strength": 9,
}
```

Wrapping the `array` here into a dictionary instructs the `dataconverter` to store a lossless compressed version of the same dataset.
The dictionary has one mandatory keyword `compress`. An additional, optional keyword `strength`, exists with which to overwrite the
default compression strength to trade-off processing time with file size reduction at the granularity of an individual dataset.

> Using compression internally forces the HDF5 library to use a different, a so-called chunked data storage layout.

Reviewer: ```suggestion
Using compression internally forces the HDF5 library to use a different, so-called chunked data storage layout.
```

Note that especially for a usage in research data management systems, like NOMAD, combining both approaches, wrapping an HDF5 file
that has internally compressed datasets into a zip file is often not additionally effective unless, the HDF5 file has a considerable
number of groups and additional internal datasets where much padding bytes were added internally by the HDF5 library when writing file.

The benefit of using compression for users is that depending on the entropy of the data a substantial reduction of the HDF5 file size and thus savings
in terms of storage space and data transfer times are possible without loosing precision. Clearly, a downside of using compression is that before
any data can be accessed and worked with, e.g., in NOMAD decompressing is required. Thanks to `h5py` functionalities this happens automatically.
The chunked storage layout is useful in that it enables a selective decompression of only those portions of the dataset required.
This is an effective but advanced lever to use when implementing more effective data processing pipelines, especially when not all data are used,

> and irrespective of the research data management system or downstream applications that consume the HDF5 file.

Reviewer: ```suggestion
This provides an effective, but more advanced, mechanism for improving data processing pipelines, particularly when only subsets of the data are required, and independently of the research data management system or downstream applications that access the HDF5 file.
```
